design/design-doc-agent.md

```diff
@@ -0,0 +1,627 @@
+type: Opaque
+data:
+  kubeconfig: MWYyZDFlMmU2N2Rm...
+```
+
+In case the annotation `agent.crossplane.io/credentials-secret-name` is not
+given on `Namespace`, then the installation configuration will be used which
+would look like the following:
+
+```bash
+helm install --namespace crossplane crossplane/agent --set credentials-secret-name=my-sa-creds
+```
+
+#### Authorization
+
+##### Remote Cluster
+
+Since there will be one agent for the whole cluster, its own mounted
+`ServiceAccount` in that remote cluster needs to have read & write permissions
+for all of the following kinds in the remote cluster listed below:
+
+* `CustomResourceDefinition`
+* `Composition`
+* `InfrastructureDefinition`
+* `InfrastructurePublication`
+* `Secret`
+* All `*Requirement` types
+
+The last one is a bit tricky because the exact list of `*Requirement` types on
+kind level is not known during installation and it's not static; new published
+APIs should be available in the remote cluster dynamically. One option is to
+allow agent to grant `Role` and `RoleBinding`s to itself as it creates the
+necessary `CustomResourceDefinition`s in the remote cluster. However, an entity
+that is able to grant permissions to itself could greatly increase the security
+posture.
+
+When you zoom out and think about how the `Role` will look like, in most of the
+cases, it's something like the following:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: crossplane-agent
+  namespace: default
+rules:
+  # all kinds could be under one company/organization group
+- apiGroups: ["acmeco.org"] 
+  resources: ["*"]
+  verbs: ["*"]
+  # or there could be logical groupings for different sets of requirements
+- apiGroups: ["database.acmeco.org"]
+  resources: ["*"]
+  verbs: ["*"]
+```
+
+As you can see, it's either one group for all the new APIs or a logical group
+for each set of APIs. In both cases, the frequency of the need to add a new
+`apiGroup` is less than one would imagine thanks to the ability of allowing a
+whole group; most frequently, the platform operators will be adding new kinds to
+the existing groups.
+
+In the light of this assumption, the initial approach will be that the `Role`
+bound to the agent will be populated by a static list of the groups of the
+requirement types during the installation like shown above and if a new group is
+introduced, then an addition to this `Role` will be needed. A separate
+controller to dynamically manage the `Role` is mentioned in the [Future
+Considerations](#future-considerations) section.
+
+##### Central Cluster
+
+The `ServiceAccount` that will be created in the central cluster needs to have
+the following permissions for agent to do its operations:
+
+* Read
+  * `CustomResourceDefinition`s
+  * `InfrastructureDefinition`s
+  * `InfrastructurePublication`s
+  * `Composition`s
+  * `Secret`s in the given namespace.
+* Write
+  * `*Requirement` types that you'd like to allow in given namespace.
```

Comment:
How does the crossplane cluster know those types? Again, are those CRDs or CRs? I assumed that those are from the remote cluster.

Reply:
They are all defined in the crossplane cluster via InfrastrucutureDefinition and InfrastructurePublications. The agent makes sure both clusters have the same set of CRDs so that the sync is lossless. The actual `*Requirement`s will come from remote cluster.